rust/wal3/src/writer.rs

```diff
@@ -615,112 +615,138 @@ impl OnceLogWriter {
         Ok(log_position)
     }
 
-    #[tracing::instrument(skip(self))]
-    async fn garbage_collect(
-        &self,
-        options: &GarbageCollectionOptions,
-        keep_at_least: Option<LogPosition>,
-    ) -> Result<(), Error> {
-        self.garbage_collect_recursive(options, false, None, keep_at_least)
-            .await
-    }
-
-    async fn garbage_collect_recursive(
+    /// Perform phase 1 of garbage collection.
+    ///
+    /// Pre-condition:  manifest/MANIFEST exists.
+    ///
+    /// Post-condition:
+    /// - gc/GARBAGE exists as a non-empty file.
+    /// - snapshots created by gc/GARBAGE get created.
+    #[tracing::instrument(skip(self, options))]
+    async fn garbage_collect_phase1(
         &self,
         options: &GarbageCollectionOptions,
-        base: bool,
-        existing: Option<&ETag>,
         keep_at_least: Option<LogPosition>,
-    ) -> Result<(), Error> {
+    ) -> Result<bool, Error> {
+        self.manifest_manager.heartbeat().await?;
         let cutoff = self.garbage_collection_cutoff().await?;
         let cutoff = if let Some(keep_at_least) = keep_at_least {
             keep_at_least.min(cutoff)
         } else {
             cutoff
         };
-        self.manifest_manager.heartbeat().await?;
-        let garbage_and_e_tag =
-            match Garbage::load(&self.options.throttle_manifest, &self.storage, &self.prefix).await
-            {
-                Ok(Some((garbage, e_tag))) => Some((garbage, e_tag)),
-                Ok(None) => None,
-                Err(err) => {
-                    return Err(err);
+        let mut attempts = 0;
+        loop {
+            attempts += 1;
+            if attempts > 3 {
+                return Err(Error::LogContentionFailure);
+            }
+            let garbage_and_e_tag =
+                match Garbage::load(&self.options.throttle_manifest, &self.storage, &self.prefix)
```

Comment:
For my understanding, this contention only happens when two GC processes concurrently update the Garbage file which in theory should not happen except maybe during deploys when more than one GC service pod GCs the same collection